broker/fragment/store_azure.go

```diff
@@ -183,30 +240,96 @@ func (a *azureBackend) azureClient(ep *url.URL) (cfg AzureStoreConfig, client pi
 		return
 	}
 
-	accountName := os.Getenv("AZURE_ACCOUNT_NAME")
-	accountKey := os.Getenv("AZURE_ACCOUNT_KEY")
-	credentials, err := azblob.NewSharedKeyCredential(accountName, accountKey)
-	if err != nil {
-		return
+	var credentials azblob.Credential
+
+	if ep.Scheme == "azure" {
+		var accountName = os.Getenv("AZURE_ACCOUNT_NAME")
+		var accountKey = os.Getenv("AZURE_ACCOUNT_KEY")
+		sharedKeyCred, err := service.NewSharedKeyCredential(accountName, accountKey)
+		if err != nil {
+			return cfg, nil, err
+		}
+		serviceClient, err := service.NewClientWithSharedKeyCredential(cfg.serviceUrl(), sharedKeyCred, &service.ClientOptions{})
+		if err != nil {
+			return cfg, nil, err
+		}
+		a.svcClient = *serviceClient
+		// Create an azblob credential that we can pass to `NewPipeline`
+		credentials, err = azblob.NewSharedKeyCredential(accountName, accountKey)
+		if err != nil {
+			return cfg, nil, err
+		}
+	} else if ep.Scheme == "azure-ad" {
+		// Link to the Azure docs describing what fields are required for active directory auth
+		// https://learn.microsoft.com/en-us/azure/developer/go/azure-sdk-authentication-service-principal?tabs=azure-cli#-option-1-authenticate-with-a-secret
+		clientId := os.Getenv("AZURE_CLIENT_ID")
+		clientSecret := os.Getenv("AZURE_CLIENT_SECRET")
+
+		identityCreds, err := azidentity.NewClientSecretCredential(
+			cfg.accountTenantID,
+			clientId,
+			clientSecret,
+			&azidentity.ClientSecretCredentialOptions{
+				AdditionallyAllowedTenants: []string{cfg.accountTenantID},
+				DisableInstanceDiscovery:   true,
+			},
+		)
+		if err != nil {
+			return cfg, nil, err
+		}
+
+		serviceClient, err := service.NewClient(cfg.serviceUrl(), identityCreds, &service.ClientOptions{})
+		if err != nil {
+			return cfg, nil, err
+		}
+		a.svcClient = *serviceClient
+
+		credentials, err = getAzureStorageCredential(identityCreds, cfg.accountTenantID)
```

Comment:
The credential that we're building here -- it seems to be specific to the current `ep` ? It embeds their `accountTenantID` ?

Note that there's just _one_ `azureBackend` instance which is registered globally to handle all fragment storage requests across all `azure://` or `azure-ad://` endpoints. As such it can have client instances and/or tokens which are coupled to our own/owned azure account, but cannot have a client that's specific to one end-user account.

Ideally any end-user-specific clients could be cheaply built from first principles by parsing an endpoint / composing with our global client for our service principal. If the underlying client needs to issue a network request each time, that's probably not reasonable however. Compare with `store_s3` and `s3Client` which keeps a multi-map of clients for each configuration of endpoint and profile.

Reply:
Yuuup, totally right. The only thing that varies here is the `accountTenantID`, which is the tenant of the customer's Azure AD account. Credentials for a particular tenant tenant will let you access any storage account and container that the app has been granted access to. So I changed this to keep track of a `map[string]pipeline.Pipeline` of pipelines per tenant ID, and `(*azureBackend).azureClient(url)` will use that map to cache its clients.